File: python/src/runapi/luma/resources/modify_video.py
"""Luma modify-video resource."""

from __future__ import annotations

from typing import Any, Optional

from runapi.core import Resource, RequestOptions

from ..contract_gen import CONTRACT
from ..types import CompletedModifyVideoResponse, ModifyVideoResponse


class ModifyVideo(Resource):
    """Modify a source video with Luma."""

    ENDPOINT = "/api/v1/luma/modify_video"

    RESPONSE_CLASS = ModifyVideoResponse
    COMPLETED_RESPONSE_CLASS = CompletedModifyVideoResponse

    def run(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        """Create a task and poll until it completes."""
        task = self.create(options=options, **params)
        return self._poll_until_complete(lambda: self.get(task.id, options=options))

    def create(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        """Create a modify-video task and return immediately with an ``id``."""
        compacted = self._compact_params(params)
        self._validate_contract(CONTRACT["modify-video"], compacted)
        return self._request("post", self.ENDPOINT, body=compacted, options=options)

    def get(self, id: str, options: Optional[RequestOptions] = None) -> Any:
        """Fetch the current status of a modify-video task."""
        return self._request("get", f"{self.ENDPOINT}/{id}", options=options)

File: ruby/lib/runapi/luma/resources/modify_video.rb
# frozen_string_literal: true

module RunApi
  module Luma
    module Resources
      # Luma video modification resource.
      # Apply prompt-guided visual edits to an existing video while preserving its motion.
      class ModifyVideo
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/luma/modify_video"
        RESPONSE_CLASS = Types::ModifyVideoResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedModifyVideoResponse

        def initialize(http)
          @http = http
        end

        # Modify a video and wait until complete.
        #
        # @param params [Hash] modification parameters
        # @return [RunApi::Luma::Types::CompletedModifyVideoResponse] completed task with videos
        def run(options: nil, **params)
          task = create(options: options, **params)
          poll_until_complete { get(task.id, options: options) }
        end

        # Start a video modification task.
        #
        # @param params [Hash] modification parameters
        # @return [RunApi::Luma::Types::ModifyVideoResponse] task creation result with id
        def create(options: nil, **params)
          params = compact_params(params)
          validate_contract!(CONTRACT["modify-video"], params)
          request(:post, ENDPOINT, body: params, options: options)
        end

        # Get video modification task status by task ID.
        #
        # @param id [String] task ID
        # @return [RunApi::Luma::Types::ModifyVideoResponse] current task status
        def get(id, options: nil)
          request(:get, "#{ENDPOINT}/#{id}", options: options)
        end
      end
    end
  end
end
